feat: init policy defaults to true

eduardacoppo opened the pull request:
The `:init` policy flag requests that the connection sends the last value written to it.
This PR applies the `:init` policy using the `recommend_init` flag

Depends on:
- [ ] https://github.com/rock-core/tools-orogen/pull/23

eduardacoppo committed:
feat: use recommend_init flag to apply init policy

eduardacoppo committed:
chore: fix rubocop offense by moving code

fix Metrics/CyclomaticComplexity offense by moving the connection validation code to a separate function

eduardacoppo committed:
test: add tests for init policy

doudou commented on lib/syskit/models/port.rb:
```diff
@@ -105,19 +105,13 @@ def to_component_port
             def connect_to(in_port, policy = {})
                 out_port = to_component_port
+                if out_port.respond_to?(:recommend_init)
+                    policy[:init] = out_port.recommend_init
```
You shouldn't be setting this if `init` was given explicitly by the caller. And you really shouldn't be changing the argument (i.e. should be `policy = policy.merge(init: out_port.recommend_init)`)

doudou commented on test/models/test_port.rb:
```diff
@@ -72,6 +72,23 @@
+        it "adds 'init: true' as a default policy" do
+            policy = {}
+            flexmock(out_task_m).should_receive(:connect_ports).explicitly.once
+                                .with(in_task_m, %w[out in] => policy.merge(init: true))
```
`policy` here is empty ... `policy.merge(init: true)` is `{ init: true }`

doudou commented on lib/syskit/models/port.rb:
```diff
@@ -105,19 +105,13 @@ def to_component_port
-                            raise WrongPortConnectionTypes.new(self, in_port), "cannot connect #{out_port} to #{in_port}: types mismatch"
-                        end
-
+                        validate_connection!(out_port, in_port)
```
Why the `!` here ?

eduardacoppo replied:
This is subjective, but while reading about `!` methods in ruby, I learned that this is a convention in rails for methods that might raise exceptions (example: https://api.rubyonrails.org/classes/ActiveRecord/Persistence.html#method-i-save-21). Seemed like a practical way to indicate that the method might interrupt the normal flow of code.
But now that you mention it, maybe it doesn't make sense here, since it's not being called by the user. It does draw attention to the method, though, and that was the main idea.

doudou replied:
It's a ... long discussion in the Ruby circles. The `!` is generally meant to represent "unexpectedly dangerous methods".... Not a very easy criteria to apply. For instance, in containers, it's used to mark self-modifying versions of modification methods (e.g. Hash#merge! vs. Hash#merge)

Regardless of the overall discussion, it's best to always follow the conventions of the codebase you're modifying. Within Syskit, the `validate_*` methods all raise and none has a `!`.

doudou commented on lib/syskit/models/port.rb:
```diff
@@ -105,19 +105,13 @@ def to_component_port
             # @raise [SelfConnection]
             def connect_to(in_port, policy = {})
                 out_port = to_component_port
+                if out_port.respond_to?(:recommend_init)
```
This guard is very suspicious to me.

eduardacoppo replied:
There's a test that tries to do a connection from an input port. Without this guard, the test breaks here.
The test:
![image](https://github.com/user-attachments/assets/4c39f264-4384-43ef-af21-1f054f209b58)

doudou reviewed (commented):
Good first step.

However, I believe this won't work for ports that are exported on a composition (since they are not using this codepath). I don't think it would work in many cases related to compositions (such as e.g. a composition that has a service as child and gets its real component as an injection in profiles).

The more generic place to resolve this would be where we compute the final connection policy. https://github.com//rock-core/tools-syskit/blob/b674ba903538d229ded38c5980cfd5549535e4eb/lib/syskit/network_generation/dataflow_dynamics.rb#L525

And it will definitely not work for data readers. For data readers, you would need to work here: https://github.com//rock-core/tools-syskit/blob/6b0d92aa0c8bc62a79ce6d58dffb330abdfcceb9/lib/syskit/dynamic_port_binding.rb#L205

eduardacoppo committed:
fix/chore: policy for data readers and exported ports

and rewrite code in port model

eduardacoppo committed:
test: add tests and update existing ones

eduardacoppo committed:
chore: rewrite tests to call init_policy

doudou commented on lib/syskit/dynamic_port_binding.rb:
```diff
@@ -203,6 +203,10 @@ def initialize(
             # Method called by {Accessor} to create the accessor object from a
             # port
             def create_accessor(port)
+                if port.respond_to?(:init_policy) &&
```
Why that check ? It's an output port, it should have the policy.

eduardacoppo replied:
See https://github.com/rock-core/tools-syskit/pull/459#discussion_r1945282078

I could validate that's an output port to make it more explicit.
`unless port.input?`

doudou commented on lib/syskit/dynamic_port_binding.rb:
```diff
@@ -203,6 +203,10 @@ def initialize(
             def create_accessor(port)
+                if port.respond_to?(:init_policy) &&
+                   [true, false].include?(port.init_policy)
+                    policy.merge(init: port.init_policy)
```
This overrides an explicit `init` flag that may be given in `policy`. The explicit value (from `policy` should take precedence

doudou commented on lib/syskit/dynamic_port_binding.rb:
```diff
@@ -203,6 +203,10 @@ def initialize(
             # port
             def create_accessor(port)
+                if port.respond_to?(:init_policy) &&
+                   [true, false].include?(port.init_policy)
```
Do not check this.

The `init_policy` is a default. Use it as a default unconditionally and let the caller override it via the various `policy` arguments.

doudou commented on lib/syskit/models/port.rb:
```diff
@@ -105,19 +105,14 @@ def to_component_port
             # @raise [SelfConnection]
             def connect_to(in_port, policy = {})
                 out_port = to_component_port
+                if out_port.respond_to?(:init_policy) &&
+                    [true, false].include?(out_port.init_policy)
+                    policy = policy.merge(init: out_port.init_policy)
+                end
```
This should not be done here at all. Just do it
- in data reader creation
- in the system network generation

eduardacoppo committed:
chore: remove unneccessary code in models/port and tests

eduardacoppo committed:
fix: do not override existing init flag

Use init_policy from the port as a default for the policy
Do not override the existing init flag in the policy
`policy.fetch(:init, port.model.init_policy)` to set the init value if not already provided in the policy

eduardacoppo committed:
fix/test: allow policy merge and add tests for this

doudou commented on lib/syskit/network_generation/dataflow_dynamics.rb:
```diff
@@ -601,17 +601,21 @@ def policy_for(
                 end
 
+                source_port_m = source_port.model
+                policy ||= {}
```
You are defining the `policy` local variable here ... It is nil on definition. I.e. these two lines are actually equivalent to

```
policy = { init: source_port_m.init_policy }
```

But then you're merging that in all the policies computed later ... would make more sense to do

```
if ...
    policy = compute_reliable...
elsif
else
end

policy.merge(init: source_port_m.init_policy)
```

doudou reviewed (commented):
Right ... so turns out that the syskit dataflow comptuation code itself is binary. Either the policy is empty and it gets computed automatically, or it's not and policy_for gets called.

Let's merge this PR following that behaviour and fix it later.

eduardacoppo committed:
fix: do not merge if no init policy

jhonasiv commented on lib/syskit/network_generation/dataflow_dynamics.rb:
```diff
@@ -603,23 +603,27 @@ def policy_for(
                 end
+
+                source_port_m = source_port.model
+                unless source_port_m.init_policy.nil?
```
```suggestion
                unless source_port_m.init_policy?
```

eduardacoppo committed:
test: fix merge test
